File: python/n437_PathSum.py
# https://leetcode-cn.com/problems/path-sum-iii/
class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None

class Solution:
    def pathSum(self, root: TreeNode, sum: int) -> int:
        if root is None: return 0
        return self.rootSum(root, sum) + self.pathSum(root.left, sum) + self.pathSum(root.right, sum)

    def rootSum(self, root: TreeNode, sum: int) -> int:
        if root is None: return 0
        return (1 if sum - root.val == 0 else 0) + self.rootSum(root.left, sum-root.val) + self.rootSum(root.right, sum-root.val)
        
        
if __name__ == "__main__":
    pass         

File: cpp/n437_PathSum.cpp
// https://leetcode-cn.com/problems/path-sum-iii/
#include <iostream>
#include <vector>
#include <string>
#include <set>
#include <algorithm>
using namespace std;

struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode(int x) : val(x), left(NULL), right(NULL) {}
};

class Solution {
public:
    int pathSum(TreeNode* root, int sum) {
        if (root == NULL) return 0;
        return (rootSum(root, sum) + pathSum(root -> left, sum) + pathSum(root -> right, sum));
    }
    
    int rootSum(TreeNode* root, int sum) {
        if (root == NULL) return 0;
        int res = 0;
        int tmp = root -> val;
        if (sum == tmp) res = 1;
        return (res + rootSum(root -> left, sum - tmp) + rootSum(root -> right, sum - tmp));
    }
};

int main(){
    return 0;
}